From 85325114c7974166d26ba147b1cf6450b9453e35 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 8 Dec 2021 15:05:15 +0100
Subject: [PATCH 0529/1695] plat-stm32mp1: update PSCI ops to define the PSCI
 suspend.

Add the psci_system_suspend support for STM32MP1 platform.

Adds some update to fix dynamic PSCI and core states link
to chip ID support.
Adds some console flush to ensure all traces are properly
displayed.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 783c48ea5239ff5078e84884718d5cc2cd0a4142 plat-stm32mp1: update PSCI ops to define the PSCI suspend.]
Change-Id: I7a7b9f7f6b696abcea4e17b40192babf8aeb47ee
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/358375
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/arch/arm/plat-stm32mp1/pm/psci.c | 229 ++++++++++++++++++++++----
 1 file changed, 197 insertions(+), 32 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/pm/psci.c b/core/arch/arm/plat-stm32mp1/pm/psci.c
index a05044bfb..38b4f87fc 100644
--- a/core/arch/arm/plat-stm32mp1/pm/psci.c
+++ b/core/arch/arm/plat-stm32mp1/pm/psci.c
@@ -5,28 +5,37 @@
 
 #include <arm.h>
 #include <boot_api.h>
+#include <config.h>
 #include <console.h>
 #include <drivers/clk.h>
 #include <drivers/rstctrl.h>
+#include <drivers/stm32mp_dt_bindings.h>
 #include <drivers/stm32mp1_pmic.h>
 #include <drivers/stm32mp1_rcc.h>
 #include <drivers/stpmic1.h>
-#include <drivers/stm32mp_dt_bindings.h>
+#include <initcall.h>
 #include <io.h>
+#include <keep.h>
 #include <kernel/cache_helpers.h>
 #include <kernel/delay.h>
 #include <kernel/boot.h>
 #include <kernel/interrupt.h>
 #include <kernel/misc.h>
 #include <kernel/panic.h>
+#include <kernel/pm.h>
 #include <kernel/spinlock.h>
 #include <mm/core_memprot.h>
 #include <platform_config.h>
+#include <sm/pm.h>
 #include <sm/psci.h>
 #include <sm/std_smc.h>
 #include <stm32_util.h>
+#include <stm32mp_pm.h>
+#include <string.h>
 #include <trace.h>
+#include <util.h>
 
+#include "context.h"
 #include "power.h"
 
 #define CONSOLE_FLUSH_DELAY_MS		10
@@ -79,22 +88,6 @@ int psci_affinity_info(uint32_t affinity, uint32_t lowest_affinity_level)
 	}
 }
 
-#ifndef CFG_PM
-/* Implements stm32_pm_cpu_power_down_wfi() when CFG_PM is disable */
-void __noreturn stm32_pm_cpu_power_down_wfi(void)
-{
-	dcache_op_level1(DCACHE_OP_CLEAN);
-
-	io_write32(stm32_rcc_base() + RCC_MP_GRSTCSETR,
-		   RCC_MP_GRSTCSETR_MPUP1RST);
-
-	dsb();
-	isb();
-	wfi();
-	panic();
-}
-#endif
-
 #if CFG_TEE_CORE_NB_CORE == 1
 /*
  * Function called when a CPU is booted through the OP-TEE.
@@ -102,7 +95,7 @@ void __noreturn stm32_pm_cpu_power_down_wfi(void)
  */
 void stm32mp_register_online_cpu(void)
 {
-	assert(core_state[0] == CORE_OFF);
+	assert(core_state[0] == CORE_OFF || core_state[0] == CORE_RET);
 	core_state[0] = CORE_ON;
 }
 #else
@@ -112,15 +105,28 @@ void stm32mp_register_online_cpu(void)
 	uint32_t exceptions = lock_state_access();
 
 	if (pos == 0) {
-		assert(core_state[pos] == CORE_OFF);
+		assert((core_state[pos] == CORE_OFF) ||
+		       (core_state[pos] == CORE_RET));
 	} else {
+		if (!stm32mp_supports_second_core())
+			return;
+
 		if (core_state[pos] != CORE_AWAKE) {
 			core_state[pos] = CORE_OFF;
 			unlock_state_access(exceptions);
 			stm32_pm_cpu_power_down_wfi();
+
+			/* should not return */
 			panic();
 		}
+
+		/* Balance BKPREG clock gating */
 		clk_disable(stm32mp_rcc_clock_id_to_clk(RTCAPB));
+
+		/* Clear hold in pen flag */
+		io_write32(stm32mp_bkpreg(BCKR_CORE1_MAGIC_NUMBER),
+			   BOOT_API_A7_RESET_MAGIC_NUMBER);
+
 	}
 
 	core_state[pos] = CORE_ON;
@@ -221,33 +227,190 @@ int psci_cpu_off(void)
 
 	thread_mask_exceptions(THREAD_EXCP_ALL);
 	stm32_pm_cpu_power_down_wfi();
+
+	/* should not return */
 	panic();
 }
 #endif
 
-/* Override default psci_system_off() with platform specific sequence */
-void __noreturn psci_system_off(void)
+static int enter_cstop_suspend(unsigned int soc_mode)
 {
-	DMSG("core %u", get_core_pos());
+	int rc = 1;
+
+	if (read_isr())
+		return rc;
 
-	if (TRACE_LEVEL >= TRACE_DEBUG) {
-		console_flush();
-		mdelay(CONSOLE_FLUSH_DELAY_MS);
+	stm32_enter_cstop(soc_mode);
+
+#ifndef CFG_STM32MP1_OPTEE_IN_SYSRAM
+	stm32mp_pm_call_bl2_lp_entry(soc_mode);
+	rc = 0;
+#else
+	if (need_to_backup_cpu_context(soc_mode)) {
+		stm32_pm_cpu_power_down_wfi();
+	} else {
+		stm32_pm_cpu_wfi();
+		rc = 0;
 	}
+#endif
+
+	stm32_exit_cstop();
 
-	if (stm32mp_with_pmic()) {
-		stm32mp_get_pmic();
-		stpmic1_switch_off();
-		udelay(100);
+	return rc;
+}
+
+static int plat_suspend(uint32_t arg)
+{
+	unsigned int soc_mode = arg;
+	size_t pos = get_core_pos();
+	int rc = 1;
+
+	if (read_isr())
+		return rc;
+
+	/* No need to lock state access as CPU is alone when here */
+	assert(core_state[pos] == CORE_ON);
+	core_state[pos] = CORE_RET;
+
+	if (stm32mp_pm_save_context(soc_mode) == TEE_SUCCESS)
+		rc = enter_cstop_suspend(soc_mode);
+
+	stm32mp_pm_restore_context(soc_mode);
+	stm32mp_pm_wipe_context();
+
+	assert(core_state[pos] == CORE_RET);
+	core_state[pos] = CORE_ON;
+
+	return rc;
+}
+
+static void plat_resume(uint32_t arg)
+{
+	unsigned int soc_mode = arg;
+
+	stm32mp_register_online_cpu();
+
+	assert(core_state[get_core_pos()] == CORE_ON);
+
+	stm32mp_pm_restore_context(soc_mode);
+}
+
+static bool plat_can_suspend(void)
+{
+	size_t pos = get_core_pos();
+	size_t n = 0;
+	uint32_t exceptions = 0;
+	bool rc = true;
+
+#ifndef CFG_STM32MP13
+	if (!IS_ENABLED(CFG_STM32_RNG))
+		return false;
+#endif
+
+	if (CFG_TEE_CORE_NB_CORE == 1)
+		return true;
+
+	exceptions = lock_state_access();
+
+	for (n = 0; n < ARRAY_SIZE(core_state); n++) {
+		if (n == pos)
+			continue;
+
+		if (core_state[n] == CORE_AWAKE) {
+			/* State core as lost and proceed suspend */
+			core_state[n] = CORE_OFF;
+		}
+
+		if (core_state[n] != CORE_OFF)
+			rc = false;
 	}
 
-	panic();
+	unlock_state_access(exceptions);
+
+	return rc;
+}
+
+/* Override default psci_system_suspend() with platform specific sequence */
+int psci_system_suspend(uintptr_t entry, uint32_t context_id __unused,
+			struct sm_nsec_ctx *nsec)
+{
+	int ret = PSCI_RET_INVALID_PARAMETERS;
+	uint32_t soc_mode = 0;
+	int __maybe_unused pos = get_core_pos();
+
+	DMSG("core %u", pos);
+
+	if (!plat_can_suspend())
+		return PSCI_RET_DENIED;
+
+	soc_mode = stm32mp1_get_lp_soc_mode(PSCI_MODE_SYSTEM_SUSPEND);
+
+	DMSG("Soc mode requested is %"PRIu32, soc_mode);
+
+	if (soc_mode == STM32_PM_CSLEEP_RUN) {
+		stm32_enter_csleep();
+		nsec->mon_lr = (uint32_t)entry;
+		return PSCI_RET_SUCCESS;
+	}
+
+	assert(cpu_mmu_enabled() && core_state[pos] == CORE_ON);
+
+	if (need_to_backup_cpu_context(soc_mode)) {
+#ifdef CFG_STM32MP1_OPTEE_IN_SYSRAM
+		if (!stm32mp_supports_hw_cryp())
+			return PSCI_RET_DENIED;
+#endif
+
+		sm_save_unbanked_regs(&nsec->ub_regs);
+		/*
+		 * sm_pm_cpu_suspend(arg, func) saves the CPU core context in
+		 * TEE RAM then calls func(arg) to run the platform lower power
+		 * sequence.
+		 *
+		 * If platform fails to suspend, sm_pm_cpu_suspend() returns
+		 * with a non null return code. When sm_pm_cpu_suspend()
+		 * returns 0 platform context must be restored.
+		 */
+		ret = sm_pm_cpu_suspend((uint32_t)soc_mode, plat_suspend);
+		if (ret == 0) {
+#ifndef CFG_STM32MP1_OPTEE_IN_SYSRAM
+			stm32_exit_cstop();
+#endif
+			plat_resume((uint32_t)soc_mode);
+			sm_restore_unbanked_regs(&nsec->ub_regs);
+		}
+	} else {
+		ret = plat_suspend((uint32_t)soc_mode);
+	}
+
+	if (ret == 0) {
+		nsec->mon_lr = (uint32_t)entry;
+		IMSG("Resumed");
+		return PSCI_RET_SUCCESS;
+	}
+
+	return PSCI_RET_INTERNAL_FAILURE;
+}
+
+/* Override default psci_system_off() with platform specific sequence */
+void __noreturn psci_system_off(void)
+{
+	uint32_t soc_mode = stm32mp1_get_lp_soc_mode(PSCI_MODE_SYSTEM_OFF);
+
+	DMSG("core %u", get_core_pos());
+
+	if (pm_change_state(PM_OP_SUSPEND, soc_mode))
+		panic();
+
+	stm32_enter_cstop_shutdown(soc_mode);
 }
 
 /* Override default psci_system_reset() with platform specific sequence */
 void __noreturn psci_system_reset(void)
 {
-	DMSG("core %u", get_core_pos());
+	IMSG("Forced system reset");
+	console_flush();
+	mdelay(10);
 	stm32_reset_system();
 	udelay(100);
 	panic();
@@ -264,13 +427,15 @@ int psci_features(uint32_t psci_fid)
 		return PSCI_RET_SUCCESS;
 	case PSCI_CPU_ON:
 	case PSCI_CPU_OFF:
-		if (CFG_TEE_CORE_NB_CORE > 1)
+		if (stm32mp_supports_second_core())
 			return PSCI_RET_SUCCESS;
 		return PSCI_RET_NOT_SUPPORTED;
 	case PSCI_SYSTEM_OFF:
 		if (stm32mp_with_pmic())
 			return PSCI_RET_SUCCESS;
 		return PSCI_RET_NOT_SUPPORTED;
+	case PSCI_SYSTEM_SUSPEND:
+		return PSCI_RET_SUCCESS;
 	default:
 		return PSCI_RET_NOT_SUPPORTED;
 	}
-- 
2.47.2

